### PR TITLE
Revert State Changes from Latest Block

### PR DESCRIPTION
**Pull request by AdityaSripal**
This PR aims to add a command to revert the latest block. This will undo any changes in Tendermint State and block store from the last time `ApplyBlock` was called. It should also rollback ApplicationState changes though that is currently unimplemented since further discussion is necessary on how that should be accomplished.

Reasons for why reverting a block would be useful are explained here: https://github.com/tendermint/tendermint/issues/3196 and https://github.com/tendermint/tendermint/issues/3845

Prior work: https://github.com/tendermint/tendermint/pull/3332

* [ ] Referenced an issue explaining the need for the change
* [ ] Updated all relevant documentation in docs
* [ ] Updated all code comments where relevant
* [ ] Wrote tests
* [ ] Updated CHANGELOG_PENDING.md

### CODE REVIEW

File: store/store.go [hunk 1]
```diff
@@ -195,6 +195,46 @@ func (bs *BlockStore) saveBlockPart(height int64, index int, part *types.Part) {
...
+	blockID := blockMeta.BlockID
+	bs.db.Delete(calcBlockMetaKey(latest))
+
+	// Delete block parts
+	for i := 0; i < blockID.PartsHeader.Total; i++ {
+		bs.db.Delete(calcBlockPartKey(latest, i))
+	}
+
+	// Delete block commit
+	bs.db.Delete(calcBlockCommitKey(latest - 1))
```
Comment (AdityaSripal) [inline store/store.go]: Not sure if I should be deleting the `SeenCommit` for the same height as well

File: state/execution.go [hunk 2]
```diff
@@ -179,6 +179,25 @@ func (blockExec *BlockExecutor) ApplyBlock(state State, blockID types.BlockID, b
 	return state, nil
 }
 
+// RevertBlock takes in a reverted block and the new latest block and returns the state
+// to its previous State before reverted block got applied
+// reverted block must be the LastBlockHeight of current state or will return error
+// Will not revert changes to mempool or evpool
+func (blockExec *BlockExecutor) RevertBlock(state State, reverted, newHead *types.Block) (State, error) {
```
Comment (AdityaSripal) [inline state/execution.go]: How should `AppState` get rolled back? Should Tendermint call some ABCI Method to request proxy connection to rollback `AppState`?

Comment (alexanderbez) [reply state/execution.go]: Great question. Maybe through ABCI we have the app `DeleteVersion` on every version up until the rollback height?

**Comment by marbar3778** [comment]
@AdityaSripal can you confirm that this can be closed. I believe I remember you mentioning that it wasn't needed anymore.

**Comment by AdityaSripal** [comment]
Yes, can you confirm @ebuchman ?

**Comment by melekes** [comment]
How's this not an issue anymore?

**Comment by melekes** [comment]
"I would double check the use case. The last time I weighed in I think we realized there was a much simpler way to do this: https://github.com/tendermint/tendermint/issues/3845#issuecomment-528388207. In any case, it may still be a useful dev tool to be able to undo blocks, but then it should be an independent binary: https://github.com/tendermint/tendermint/issues/3196#issuecomment-459872337" - @ebuchman

**Comment by tessr** [comment]
Howdy folks. I'm going to close this PR--it looks like the motivations for it are captured elsewhere as well, and it sounds like this isn't the implementation strategy we'd want at this point.

**Pull request closed without merging**